File: 프로그래머스/동영상 재생기.java
package 프로그래머스;

public class Solution {
    private int totalVideoRuntime; 
    private int opStartTime;
    private int opEndTime; 

    public String solution(String video_len, String pos, String op_start, String op_end, String[] commands) {

        int time = convertTimeToSeconds(pos);
        totalVideoRuntime = convertTimeToSeconds(video_len); 

        opStartTime = convertTimeToSeconds(op_start); 
        opEndTime = convertTimeToSeconds(op_end);

        for (String command : commands) {
            time = handleButton(command, time); 
        }

        if (isInOpening(time)) {
            time = opEndTime;
        }

        String answer = convertSecondsToTime(time);
        return answer;
    }

    public int convertTimeToSeconds(String time) {
        String[] splitted_time = time.split(":");
        int min = Integer.parseInt(splitted_time[0]); 
        int sec = Integer.parseInt(splitted_time[1]);

        return sec + min * 60; 
    }

    public String convertSecondsToTime(int seconds) {
        int min = seconds / 60; 
        int sec = seconds % 60;

        String minStr;
        if (min < 10) {
            minStr = "0" + String.valueOf(min);
        } else {
            minStr = String.valueOf(min);
        }

        String secStr;
        if (sec < 10) {
            secStr = "0" + String.valueOf(sec); 
        } else {
            secStr = String.valueOf(sec); 
        }

        return minStr + ":" + secStr;
    }

    public int handleButton(String command, int currPos) {
        if (isInOpening(currPos)) {
            currPos = opEndTime;
        }

        switch (command) {
            case "prev": 
                return goBack(currPos); 
            case "next": 
                return goFront(currPos); 
            default: 
                return currPos;
        }
    }

    public int goBack(int currPos) {
        int newPos = currPos - 10;
        if (newPos < 0) {
            newPos = 0; 
        }
        return newPos;
    }

    public int goFront(int currPos) {
        int newPos = currPos + 10;
        if (totalVideoRuntime - newPos < 10) {
            newPos = totalVideoRuntime;
        }
        return newPos;
    }

    public boolean isInOpening(int currPos) {
        if (opStartTime <= currPos && currPos <= opEndTime) {
            return true;
        } 
        return false;
    }
}

File: javascript/동영상 재생기.js
function solution(video_len, pos, op_start, op_end, commands) {
  var curr = timeToSeconds(pos);
  var opStart = timeToSeconds(op_start);
  var opEnd = timeToSeconds(op_end);
  var videoEnd = timeToSeconds(video_len);

  // 처음 위치가 오프닝 구간일 경우를 고려해 jump 수행
  curr = jumpFunc(curr, opStart, opEnd);

  commands.forEach((command) => {
    switch (command) {
      case "next":
        curr = nextFunc(curr, videoEnd);
        break;
      case "prev":
        curr = prevFunc(curr);
        break;
      default:
        break;
    }
    curr = jumpFunc(curr, opStart, opEnd);
  });

  return secondsToTime(curr);
}

function prevFunc(curr) {
  if (curr < 10) {
    curr = 0;
  } else {
    curr -= 10;
  }

  return curr;
}

function nextFunc(curr, videoEnd) {
  if (curr + 10 > videoEnd) {
    curr = videoEnd;
  } else {
    curr += 10;
  }

  return curr;
}

function jumpFunc(curr, opStart, opEnd) {
  if (opStart <= curr && curr <= opEnd) {
    curr = opEnd;
  }

  return curr;
}

function timeToSeconds(time) {
  const timeSplit = time.split(":");
  let min = parseInt(timeSplit[0], 10);
  let sec = parseInt(timeSplit[1], 10);

  return sec + min * 60;
}

function secondsToTime(time) {
  var min = Math.floor(time / 60);
  var sec = time % 60;

  var minStr = String(min).padStart(2, "0");
  var secStr = String(sec).padStart(2, "0");

  return `${minStr}:${secStr}`;
}
